### PR TITLE
[WIP] dvcignore: update description

### LINKED ISSUE #485: update dvcignore docs
With merging of https://github.com/iterative/dvc/pull/2161
dvcignored is supported in every walk in the code. 
Reflect that in docs.

### PR DESCRIPTION
**Pull request by pared**
**IMPORTANT NOTES** (please read, then delete):

- Have you followed the guidelines in our
  [Contributing Documentation](https://dvc.org/doc/user-guide/contributing-documentation)?

- The PR title should start with "Fix #bugnum: " (if applicable), followed by a
  clear one-line present-tense summary of the changes introduced in the PR. For
  example: "Fix #bugnum: Introduce the first version of the collection editor.".

- Please make sure to mention "Fix #bugnum" (if applicable) somewhere in the
  description of the PR. This enables Github to link the PR to the corresponding
  bug.

Fixes #485

### CODE REVIEW

File: static/docs/user-guide/dvcignore.md [hunk 1]
```diff
@@ -18,21 +17,22 @@ provide similar functionality as `.gitignore` files provide for `git`.
...
-gradually in different parts of the project.
-
-Currently `.dvcignore` files will be read and applied in any operation that
-collects DVC-files (e.g. `checkout`, `metrics`, `status`, `run`, `repro`), so it
-is advised to use it in cases described in the first paragraph, when amount of
-files in tree of repository directory causes performance issues.
 
+## Notes
+
+- `.dvcignore` is no longer applied only on operations collecting DVC-files.
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: no need to mention this. No need to reference how was it before. We need to describe the current state.

File: static/docs/user-guide/dvcignore.md [hunk 2]
```diff
@@ -18,21 +17,22 @@ provide similar functionality as `.gitignore` files provide for `git`.
...
-is advised to use it in cases described in the first paragraph, when amount of
-files in tree of repository directory causes performance issues.
 
+## Notes
+
+- `.dvcignore` is no longer applied only on operations collecting DVC-files. 
+  Each time DVC traverses directory, dvcignore files up to project root tree 
+  are respected. 
+- ignoredd files will not be cached and preserved in any way, they will 
+  be non-existent for DVC. Its worth to remember that, especially when ignoring
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: Its -> It's

File: static/docs/user-guide/dvcignore.md [hunk 3]
```diff
@@ -18,21 +17,22 @@ provide similar functionality as `.gitignore` files provide for `git`.
...
-collects DVC-files (e.g. `checkout`, `metrics`, `status`, `run`, `repro`), so it
-is advised to use it in cases described in the first paragraph, when amount of
-files in tree of repository directory causes performance issues.
 
+## Notes
+
+- `.dvcignore` is no longer applied only on operations collecting DVC-files. 
+  Each time DVC traverses directory, dvcignore files up to project root tree 
+  are respected. 
+- ignoredd files will not be cached and preserved in any way, they will
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: ignored

Comment (shcheklein) [reply static/docs/user-guide/dvcignore.md]: it needs more details on what does "preserved" means. It's not clear. For example, when we remove outputs do we skip ignored files or remove them as well? It worth describing and mentioning these edge cases. And give examples below.

File: static/docs/user-guide/dvcignore.md [hunk 4]
```diff
@@ -18,21 +17,22 @@ provide similar functionality as `.gitignore` files provide for `git`.
...
-might be tricky to handle (e.g. what to do when we are `dvc add`-ing directory
-containing `.dvcignore` file). Therefore, we decided to enable this feature
-gradually in different parts of the project.
-
-Currently `.dvcignore` files will be read and applied in any operation that
-collects DVC-files (e.g. `checkout`, `metrics`, `status`, `run`, `repro`), so it
-is advised to use it in cases described in the first paragraph, when amount of
-files in tree of repository directory causes performance issues.
 
+## Notes
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: No need to put Node here. It should be now integral part of the description. May be some phrase similar to `dvc add` that describes steps, what's going on internally, etc.

File: static/docs/user-guide/dvcignore.md [hunk 5]
```diff
@@ -18,21 +17,22 @@ provide similar functionality as `.gitignore` files provide for `git`.
...
+
+- `.dvcignore` is no longer applied only on operations collecting DVC-files. 
+  Each time DVC traverses directory, dvcignore files up to project root tree 
+  are respected. 
+- ignoredd files will not be cached and preserved in any way, they will 
+  be non-existent for DVC. Its worth to remember that, especially when ignoring
+  files inside DVC-handled directories.
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: use capital letter to start full sentences in a bulleted list

File: static/docs/user-guide/dvcignore.md [hunk 6]
```diff
@@ -18,21 +17,22 @@ provide similar functionality as `.gitignore` files provide for `git`.
...
+- ignoring files inside directory dependencies under DVC control means that
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: hard to understand. It should be something like . It should be simply "Keep in mind, that after you added some files to .dvcignore that affects one of the existing outputs ..."

File: static/docs/user-guide/dvcignore.md [hunk 7]
```diff
@@ -18,21 +17,22 @@ provide similar functionality as `.gitignore` files provide for `git`.
...
+## Notes
+
+- `.dvcignore` is no longer applied only on operations collecting DVC-files. 
+  Each time DVC traverses directory, dvcignore files up to project root tree 
+  are respected. 
+- ignoredd files will not be cached and preserved in any way, they will 
+  be non-existent for DVC. Its worth to remember that, especially when ignoring
+  files inside DVC-handled directories.
+- ignoring files inside directory dependencies under DVC control means that 
+  this dependency has changed, as if ignored files were deleted, so commiting
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: is it about dependency or outputs as well?

File: static/docs/user-guide/dvcignore.md [hunk 8]
```diff
@@ -18,21 +17,22 @@ provide similar functionality as `.gitignore` files provide for `git`.
...
+  "new" version of dependency might be required.
+- If DVC will stumble upon `.dvcignore` file inside dependency directory, it
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: is it about dependency or outputs as well?

File: static/docs/user-guide/dvcignore.md (hunk 3 above)
Comment (pared) [reply static/docs/user-guide/dvcignore.md]: I agree, I think I should remve the preserved part. The most important thinkg about this point is that ignored files will not be cached. Ill inkclude removing edge case in new point.

File: static/docs/user-guide/dvcignore.md (hunk 7 above)
Comment (pared) [reply static/docs/user-guide/dvcignore.md]: Outputs too.

File: static/docs/user-guide/dvcignore.md [hunk 9]
```diff
@@ -7,32 +7,26 @@ Sometimes you might want DVC to ignore files while traversing the project
...
-files handling. When fully implemented, their implementation is intended to
-provide similar functionality as `.gitignore` files provide for `git`.
+files handling. `.dvcignore` by design works similar way as `.gitignore` does.
 
 ## How does it work
 
 - You need to create `.dvcignore` file.
 - Populate it with [patterns](https://git-scm.com/docs/gitignore) that you would
-  like to ignore.
+  like to ignore
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: why this change?

File: static/docs/user-guide/dvcignore.md [hunk 10]
```diff
@@ -7,32 +7,26 @@ Sometimes you might want DVC to ignore files while traversing the project
...
-
-Currently `.dvcignore` files will be read and applied in any operation that
-collects DVC-files (e.g. `checkout`, `metrics`, `status`, `run`, `repro`), so it
-is advised to use it in cases described in the first paragraph, when amount of
-files in tree of repository directory causes performance issues.
-
+- Ignored files will not be cached, they will be non-existent for DVC. It's
+  worth to remember that, especially when ignoring files inside DVC-handled
+  directories.
+- Keep in mind, that when you add to dvcignore entries that affect one of the
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: dvcignore -> `.dvcignore` (please, use the same terminology across the docs)

File: static/docs/user-guide/dvcignore.md [hunk 11]
```diff
@@ -7,32 +7,26 @@ Sometimes you might want DVC to ignore files while traversing the project
...
-might be tricky to handle (e.g. what to do when we are `dvc add`-ing directory
-containing `.dvcignore` file). Therefore, we decided to enable this feature
-gradually in different parts of the project.
-
-Currently `.dvcignore` files will be read and applied in any operation that
-collects DVC-files (e.g. `checkout`, `metrics`, `status`, `run`, `repro`), so it
-is advised to use it in cases described in the first paragraph, when amount of
-files in tree of repository directory causes performance issues.
-
+- Ignored files will not be cached, they will be non-existent for DVC. It's
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: `cached` -> taken under DVC control, or saved. `cached` is an internal term more or less. It's always better to explain stuff in terms that come from the end-user world, not from the internal implementation.

File: static/docs/user-guide/dvcignore.md [hunk 12]
```diff
@@ -7,32 +7,26 @@ Sometimes you might want DVC to ignore files while traversing the project
...
+  worth to remember that, especially when ignoring files inside DVC-handled
+  directories.
+- Keep in mind, that when you add to dvcignore entries that affect one of the
+  existing outputs, its status will change and DVC will behave as if that
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: again, better to use terms like `existing files or directories (outputs) that are controlled by DVC, corresponding DVC-files status ...`

Or alternatively you can use <abbr>outputs</abbr> and add new term to the glossary.js file. It's quite easy :)

File: static/docs/user-guide/dvcignore.md [hunk 13]
```diff
@@ -7,32 +7,26 @@ Sometimes you might want DVC to ignore files while traversing the project
...
+  affected files were deleted.
+- If DVC will stumble upon `.dvcignore` file inside dependency or output
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: If DVC stumbles upon .... it raises an error.

File: static/docs/user-guide/dvcignore.md (hunk 11 above)
Comment (pared) [reply static/docs/user-guide/dvcignore.md]: Lets meet halfway: saved in cache

**Comment by pared** [comment]
> Good stuff. Check my comments and would be great to put more examples?

Yes, Ill introduce them.

File: src/Documentation/glossary.js [hunk 14]
```diff
@@ -29,6 +29,13 @@ export default {
         'result (such as extracted features or a ML model file) that is ' +
         'under DVC control. Refer to [Data and Model Files Versioning]' +
         '(/doc/use-cases/data-and-model-files-versioning) for more details.'
+    },
+    {
+      name: 'Output',
+      match: ['output', 'outputs'],
+      desc:
+        'File or directory that is a result of `run` command.'
```
Comment (shcheklein) [inline src/Documentation/glossary.js]: this not exactly true. It can be import, dvc add, import-url also. It's good to mention these commands, but it's better to define it some user friendly way. What does output from a user perspective mean?

File: static/docs/user-guide/dvcignore.md (hunk 12 above)
Comment (shcheklein) [reply static/docs/user-guide/dvcignore.md]: you need to put `<abbr>` around outputs to make it work

File: static/docs/user-guide/dvcignore.md [hunk 15]
```diff
@@ -13,17 +13,17 @@ files handling. `.dvcignore` by design works similar way as `.gitignore` does.
...
 - During execution of commands that traverse directories, DVC will ignore
   matching paths.
-- Ignored files will not be cached, they will be non-existent for DVC. It's
-  worth to remember that, especially when ignoring files inside DVC-handled
-  directories.
-- Keep in mind, that when you add to dvcignore entries that affect one of the
+- Ignored files will not be saved in cache, they will be non-existent for DVC.
+  It's worth to remember that, especially when ignoring files inside 
+  DVC-handled directories.
+- Keep in mind, that when you add to .dvcignore entries that affect one of the
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: `.dvcignore` - please use the same style everywhere.

File: static/docs/user-guide/dvcignore.md [hunk 16]
```diff
@@ -13,17 +13,17 @@ files handling. `.dvcignore` by design works similar way as `.gitignore` does.
...
-  like to ignore
+  like to ignore.
 - Each line should contain only one pattern.
 - During execution of commands that traverse directories, DVC will ignore
   matching paths.
-- Ignored files will not be cached, they will be non-existent for DVC. It's
-  worth to remember that, especially when ignoring files inside DVC-handled
-  directories.
-- Keep in mind, that when you add to dvcignore entries that affect one of the
+- Ignored files will not be saved in cache, they will be non-existent for DVC.
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: we should make it clear with some note that DVC removes ignored files on `dvc repro`, `dvc run`, etc. It's super important to understand I think. And it's not clear from the description now.

File: static/docs/user-guide/dvcignore.md [hunk 17]
```diff
@@ -13,17 +13,17 @@ files handling. `.dvcignore` by design works similar way as `.gitignore` does.
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: we need to update the paragraph above, it's too tailored to the "traversing" case

File: static/docs/user-guide/dvcignore.md (hunk 16 above)
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: I would also split basic "create file, add item, profit" part from the more descriptive items - like how it works. It can be two different sections.

File: static/docs/user-guide/dvcignore.md [hunk 18]
```diff
@@ -38,7 +38,107 @@ by design works similar way as `.gitignore` does.
 
 The same as for [`.gitignore`](https://git-scm.com/docs/gitignore).
 
-## Example
+## Examples: Modification of ignored data
+
+Lets see if what happens when we modify ignored file.
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: `if what` - typo

File: static/docs/user-guide/dvcignore.md [hunk 19]
```diff
@@ -38,7 +38,107 @@ by design works similar way as `.gitignore` does.
...
+$ echo data2 >> data/data2
+$ tree .
+
+.
+└── data
+    ├── data1
+    └── data2
+```
+
+We created data directory. Lets ignore part of data and add `data` under DVC
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: use `data` everywhere as a code block. I also think you need to use `the` - created _the_ data directory. @jorgeorpinel can confirm that :)

File: static/docs/user-guide/dvcignore.md [hunk 20]
```diff
@@ -38,7 +38,107 @@ by design works similar way as `.gitignore` does.
...
+├── 54
+│   └── 40cb5e4c57ab54af68127492334a23.dir
+└── ed
+    └── c3d3797971f12c7f5e1d106dd5cee2
+```
+
+As we can see, `data1` has been ignored. Cache contains only one file entry
+(for `data2`) and one dir entry (`data`).
+
+Now, lets modify data1 and see if it affects `dvc status`.
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: again `data1`

File: src/Documentation/glossary.js [hunk 21]
```diff
@@ -37,6 +37,15 @@ export default {
       desc:
         'Stage (DVC-file) created with the `dvc import` or `dvc import-url` ' +
         'commands. They represent files or directories from external sources.'
+    },
+    {
+      name: 'Output',
+      match: ['output', 'outputs'],
+      desc:
+        'File or directory that should be handled by DVC and stored using ' +
```
Comment (shcheklein) [inline src/Documentation/glossary.js]: let's simplify it:

```
A file or a directory that is under DVC control. See `dvc add`, `dvc run`, `dvc import`, `dvc import-url` commands.
```

File: static/docs/user-guide/dvcignore.md [hunk 22]
```diff
@@ -1,14 +1,15 @@
...
-Sometimes you might want DVC to ignore files while traversing the project
-directory. For example, when working on a project with many files in its data
+Sometimes you might want DVC to ignore some files while working with the
+project. For example, when working on a project with many files in its data
 directory, you might encounter extended execution time for operations that are
-as simple as `dvc status`. To prevent this, we are implementing `.dvcignore`
-files handling. When fully implemented, their implementation is intended to
-provide similar functionality as `.gitignore` files provide for `git`.
+as simple as `dvc status`. In other case you might want to omit files or
+folders unrelated to the project (like `.DS_Store` on Mac). To address this
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: this -> these

File: static/docs/user-guide/dvcignore.md [hunk 23]
```diff
@@ -18,26 +19,126 @@ provide similar functionality as `.gitignore` files provide for `git`.
 - Each line should contain only one pattern;
 - During execution of commands that traverse directories, DVC will ignore
   matching paths;
-- Not every operation supports `.dvcignore`. To see current limitations, read
-  following paragraph.
+  
+## Remarks
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: a space after the title

File: static/docs/user-guide/dvcignore.md [hunk 24]
```diff
@@ -18,26 +19,126 @@ provide similar functionality as `.gitignore` files provide for `git`.
 - Each line should contain only one pattern;
 - During execution of commands that traverse directories, DVC will ignore
   matching paths;
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: traverse directories - it's not only about dirs, right?

File: static/docs/user-guide/dvcignore.md [hunk 25]
```diff
@@ -18,26 +19,126 @@ provide similar functionality as `.gitignore` files provide for `git`.
 - Each line should contain only one pattern;
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: please, use `.` instead of `;` at the end of each item

File: static/docs/user-guide/dvcignore.md [hunk 26]
```diff
@@ -18,26 +19,126 @@ provide similar functionality as `.gitignore` files provide for `git`.
...
   matching paths;
-- Not every operation supports `.dvcignore`. To see current limitations, read
-  following paragraph.
+  
+## Remarks
+- Ignored files will not be saved in cache, they will be non-existent for DVC.
+  It's worth to remember that, especially when ignoring files inside 
+  DVC-handled directories. **It is crucial to understand, that DVC might remove
+  ignored files upon `dvc run` or `dvc repro`. If they are not produced by
+  [pipeline](/doc/get-started/pipeline) step, they can be deleted permanently.**
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: by a pipeline

File: static/docs/user-guide/dvcignore.md [hunk 27]
```diff
@@ -18,26 +19,126 @@ provide similar functionality as `.gitignore` files provide for `git`.
...
+- Keep in mind, that when you add to .dvcignore entries that affect one of the
+  existing <abbr>outputs</abbr>, its status will change and DVC will behave
+  as if that affected files were deleted;
+- If DVC stumbles upon `.dvcignore` file inside dependency or
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: a dependency or an output

File: static/docs/user-guide/dvcignore.md [hunk 28]
```diff
@@ -18,26 +19,126 @@ provide similar functionality as `.gitignore` files provide for `git`.
...
+  <abbr>output</abbr> directory, it will raise an error. Ignoring files
```
Comment (shcheklein) [inline static/docs/user-guide/dvcignore.md]: I believe it should be `it raises`

File: static/docs/user-guide/dvcignore.md (hunk 24 above)
Comment (pared) [reply static/docs/user-guide/dvcignore.md]: Ignore is not only about dirs, but it works only during traversing of dirs :)

File: static/docs/user-guide/dvcignore.md (hunk 25 above)
Comment (pared) [reply static/docs/user-guide/dvcignore.md]: Ok, but I did it to match changes introduced for unification of bullet lists. Isn't it the way we are handling bullet lists now?

**Comment by shcheklein** [comment]
thanks!

File: static/docs/user-guide/dvcignore.md (hunk 18 above)
Comment (jorgeorpinel) [reply static/docs/user-guide/dvcignore.md]: Also `Lets`->`Let's`. I fixed both in c8591d4

File: static/docs/user-guide/dvcignore.md (hunk 20 above)
Comment (jorgeorpinel) [reply static/docs/user-guide/dvcignore.md]: Also `lets` -> `let's`. Fixed in c8591d4

File: static/docs/user-guide/dvcignore.md (hunk 25 above)
Comment (jorgeorpinel) [reply static/docs/user-guide/dvcignore.md]: Nope, we're avoiding `;` moving forward (: Refer to https://github.com/iterative/dvc.org/issues/491#issuecomment-514118621

File: static/docs/user-guide/dvcignore.md (hunk 27 above)
Comment (jorgeorpinel) [reply static/docs/user-guide/dvcignore.md]: Also not sure about phrase `stumbles upon`. Changed to `finds a` in c0aa902
